url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/methodology/History.java

```diff
@@ -0,0 +1,380 @@
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.ObjectOutputStream;
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+
+import protocols.specifications.DBProtocol;
+
+import utils.Latex;
+import attributes.Attribute;
+import attributes.CryptoCalls;
+import attributes.DistanceFraudProbability;
+import attributes.FinalSlowPhase;
+import attributes.MafiaFraudProbability;
+import attributes.Memory;
+import attributes.NumberOfRounds;
+import attributes.SizeOfMessages;
+import attributes.TerroristFraudProbability;
+import attributes.YearOfPublication;
+import attributes.relations.FinalSlowPhaseRelation;
+import attributes.relations.IntegerRelation;
+import attributes.relations.MemoryRelation;
+import attributes.relations.ProbabilityRelation;
+import attributes.relations.SizeOfMessagesRelation;
+import attributes.scales.KbitsScale;
+import attributes.scales.LogScale;
+import attributes.scales.NoScale;
+import methodology.ParetoFrontier;
+
+/*Trujillo- Apr 16, 2014
+ * Since this class considers the history of the DB protocols, it normally uses the attribute year.*/
+public class History {
+
+
+	public static void main(String[] args) throws IOException {
+		
+		System.setOut(new PrintStream("history.txt"));
+		DBProtocol[][] protocols = DBProtocol.loadProtocolsFairly();
+		Attribute[] attributes = new Attribute[]{
+				new MafiaFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new DistanceFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new TerroristFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new NumberOfRounds(new IntegerRelation(), new NoScale<Integer>()),
+				new SizeOfMessages(new SizeOfMessagesRelation(), new NoScale<Integer>()),
+				new CryptoCalls(new IntegerRelation(), new NoScale<Integer>()),
+				new Memory(new MemoryRelation(), new KbitsScale()),
+				new FinalSlowPhase(new FinalSlowPhaseRelation(), new NoScale<Boolean>()),
+				new YearOfPublication(new IntegerRelation(), new NoScale<Integer>()),
+		};
+		
+		ParetoFrontier[] frontiers = ParetoFrontier.computeAllParetoFrontiers(protocols, attributes);
+		
+		saveInDiskTheFrontiers(frontiers, "history.obj");
+		
+		
+		//printLatexTable(protocols, attributes, frontiers, 0, 8, "history_table.tex");
+		//printHistory(protocols, attributes, frontiers);
+
+		//TreeBasedProtocol tree = new TreeBasedProtocol();
+		//System.out.println(tree.getMemory(63));
+		//KimAndAvoineProtocol kim = new KimAndAvoineProtocol(0.25, 128);
+		//DominatingRelation relation = new DominatingRelation(order, protocols, attributes, 64);
+		//DBProtocol[] protocols = new DBProtocol[]{tree, kim};
+		//relation.printInfoOfDomination(tree, kim, 63);
+	}
+
+	public static void saveInDiskTheFrontiers(ParetoFrontier[] frontiers,
+			String file) throws FileNotFoundException, IOException {
+		ObjectOutputStream out = new ObjectOutputStream(new FileOutputStream(file));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at lines 75, 76. The resource is closed at line 77. There are other execution paths that do not contain closure statements, for example, when ObjectOutputStream.writeObject throws an exception. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).